src/grid.ts

```diff
@@ -1,11 +1,331 @@
+  { x: 1, y: -1 }, // NE
+  { x: 1, y: 1 }, // SE
+  { x: -1, y: 1 }, // SW
+  { x: -1, y: -1 }, // NW
+];
+
+export const ALL = [...CARDINAL, ...DIAGONAL];
+
+type Dimensions = {
+  width: number;
+  height: number;
+};
+
+export const getNeighbors = (
+  pos: Pos,
+  dir: string,
+  boundary: Dimensions,
+  asIds: boolean
+): Array<Pos> | Array<string> => {
+  let direction: Points = [];
+  if (dir === "cardinal") direction = [...CARDINAL];
+  if (dir === "diagonal") direction = [...DIAGONAL];
+  if (dir === "all") direction = [...CARDINAL, ...DIAGONAL];
+
+  const neighbors: Array<Pos> = [];
+
+  for (let dir of direction) {
+    let candidate: Pos = {
+      x: pos.x + dir.x,
+      y: pos.y + dir.y,
+      z: pos.z,
+    };
+    if (
+      candidate.x >= 0 &&
+      candidate.x < boundary.width &&
+      candidate.y >= 0 &&
+      candidate.y < boundary.height
+    ) {
+      neighbors.push(candidate);
+    }
+  }
+
+  if (asIds) return neighbors.map(toPosId);
+
+  return neighbors;
+};
+
+export const isAtSamePosition = (pos1: Pos, pos2: Pos): boolean => {
+  return pos1.x === pos2.x && pos1.y === pos2.y && pos1.z === pos2.z;
+};
+
+export const isNeighbor = (pos1: Pos, pos2: Pos): boolean => {
+  if (pos1.z !== pos2.z) {
+    return false;
+  }
+
+  const { x: ax, y: ay } = pos1;
+  const { x: bx, y: by } = pos2;
+
+  if (
+    (ax - bx === 1 && ay - by === 0) ||
+    (ax - bx === 0 && ay - by === -1) ||
+    (ax - bx === -1 && ay - by === 0) ||
+    (ax - bx === 0 && ay - by === 1)
+  ) {
+    return true;
+  }
+
+  return false;
+};
+
+export const randomNeighbor = (pos: Pos): Pos => {
+  const direction = _.sample(CARDINAL);
+  const x = pos.x + direction!.x;
+  const y = pos.y + direction!.y;
+  return { x, y, z: pos.z };
+};
+
+type DirMap = { [key: string]: number };
+
+export const getNeighbor = (pos: Pos, dir: string): Pos => {
+  const dirMap: DirMap = { N: 0, E: 1, S: 2, W: 3, NE: 4, SE: 5, SW: 6, NW: 7 };
+  const direction = ALL[dirMap[dir]];
+  return {
+    x: pos.x + direction.x,
+    y: pos.y + direction.y,
+    z: pos.z,
+  };
+};
+
+type Direction = {
+  dir: string;
+  x: number;
+  y: number;
+};
+
+export const getDirection = (posA: Pos, posB: Pos):Direction => {
+  const { x: ax, y: ay } = posA;
+  const { x: bx, y: by } = posB;
+
+  let dir = '';
+
+  if (ax - bx === -1 && ay - by === -1) dir = "NW";
+  if (ax - bx === 1 && ay - by === -1) dir = "NE";
+  if (ax - bx === 1 && ay - by === 1) dir = "SE";
+  if (ax - bx === -1 && ay - by === 1) dir = "SW";
+
+  if (ax - bx === 1 && ay - by === 0) dir = "E";
+  if (ax - bx === 0 && ay - by === -1) dir = "N";
+  if (ax - bx === -1 && ay - by === 0) dir = "W";
+  if (ax - bx === 0 && ay - by === 1) dir = "S";
+
+  if (ax - bx === 0 && ay - by === 0) dir = "X";
+
+  return { dir, x: ax - bx, y: ay - by };
+};
+
+// is this right? wonder if maybe it should be reversed...
+// you go from point a to point b... right?
+export const getVelocity = (posA:Pos, posB:Pos):Point => {
```

Comment:
if you're travelling from A to B, to compute velocity you'd do B-A. if you normalize it, you get direction (from A to B)

i usually label these `from` and `to` instead of A/B🤷‍♂️